### PR TITLE
Add snaps exports documentation

### PR DESCRIPTION
**Pull request by GuillaumeRx**
Fixes https://github.com/MetaMask/snaps-monorepo/issues/771

### CODE REVIEW

File: docs/guide/snaps-exports.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Do you have feature requests? Other ideas? We'd love to hear about them! [Click here](https://github.com/MetaMask/snaps-skunkworks/discussions) to join the discussion.
+:::
+
+## Table of Contents
+
+[[toc]]
+
+## `onRpcRequest`
+
+In order to communicate with the outside world, the snap must implement its own JSON-RPC API by exposing an exported function called `onRpcRequest`. Whenever the snap receives a JSON-RPC request from an external entity (a dapp or even another snap), this handler function will be called with the above parameters.
```
Comment (hmalik88) [inline docs/guide/snaps-exports.md:17]:
```suggestion
In order to communicate with dApps and other snaps, the snap must implement its own JSON-RPC API by exposing an exported function called `onRpcRequest`. Whenever the snap receives a JSON-RPC request, the `onRpcRequest` handler function will be called with the below parameters.
```

File: docs/guide/snaps-exports.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+```typescript
+type RpcHandlerReturn = Promise<unknown>;
+```
+
+`RpcHandlerReturn` - A promise containing the return of the implemented RPC Method.
+
+## `onTransaction`
+
+If the snap whats to provide transactions insights before they are accepted for signing by the uses, the snap must export a function called `onTransaction`. Whenever a transaction is submited to the extension this function will be called. The raw transaction payload, before it is signed, will be passed to this method.
```
Comment (hmalik88) [inline docs/guide/snaps-exports.md:44]:
```suggestion
If the snap wants to provide transaction insights before a user signs a transaction, the snap must export a function called `onTransaction`. Whenever there is a contract interaction and a transaction is submitted via the extension, this function will be called. The raw unsigned transaction payload will be passed to the `onTransaction` handler function.
```

File: docs/guide/snaps-exports.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+:::
+
+### Parameters
+
+- `RpcHandlerArgs` - The origin and the JSON-RPC request.
+
+```typescript
+interface RpcHandlerArgs = {
+  origin: string;
+  request: JsonRpcRequest<unknown[] | { [key: string]: unknown }>;
```
Comment (ritave) [inline docs/guide/snaps-exports.md:30]: `JsonRpcRequest` is not defined anywhere in the documentation

File: docs/guide/snaps-exports.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+```
+
+### Returns
+
+```typescript
+type RpcHandlerReturn = Promise<unknown>;
+```
+
+`RpcHandlerReturn` - A promise containing the return of the implemented RPC Method.
+
```
Comment (ritave) [inline docs/guide/snaps-exports.md:41]: Please add an example section. Both Typescript and Javascript

File: docs/guide/snaps-exports.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+interface RpcHandlerArgs = {
+  origin: string;
+  request: JsonRpcRequest<unknown[] | { [key: string]: unknown }>;
+};
+```
+
+### Returns
+
+```typescript
+type RpcHandlerReturn = Promise<unknown>;
```
Comment (ritave) [inline docs/guide/snaps-exports.md:37]: `Promise<unknown> | unknown`

File: docs/guide/snaps-exports.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+In order for the extension to call the `onTransaction` method of the snap, the `endowment:transaction-insight` permission must be requested. see [Permissions](./snaps-permissions.html#endowment-transaction-insight)
+:::
+
+### Parameters
+
+- `onTransactionArgs` - the raw transaction payload and the [CAIP2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) chain ID.
+
+```typescript
+interface OnTransactionArgs {
+  transaction: Record<string, unknown>;
```
Comment (ritave) [inline docs/guide/snaps-exports.md:56]: We don't explain the transaction anywhere in the documentation. We could at least direct people to [SIP-3](https://metamask.github.io/SIPs/SIPS/sip-3) apendix.

File: docs/guide/snaps-exports.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+If the snap whats to provide transactions insights before they are accepted for signing by the uses, the snap must export a function called `onTransaction`. Whenever a transaction is submited to the extension this function will be called. The raw transaction payload, before it is signed, will be passed to this method.
+
+::: tip Requesting the transaction insight permission
+In order for the extension to call the `onTransaction` method of the snap, the `endowment:transaction-insight` permission must be requested. see [Permissions](./snaps-permissions.html#endowment-transaction-insight)
+:::
+
+### Parameters
+
+- `onTransactionArgs` - the raw transaction payload and the [CAIP2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) chain ID.
```
Comment (ritave) [inline docs/guide/snaps-exports.md:52]:
```suggestion
- `onTransactionArgs` - the raw transaction payload and the [CAIP-2 chain ID](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md).
```

File: docs/guide/snaps-exports.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+interface OnTransactionArgs {
+  transaction: Record<string, unknown>;
+  chainId: string;
+}
+```
+
+### Returns
+
+```typescript
+type onTransactionHandlerReturn = Promise<onTransactionResponse>;
```
Comment (ritave) [inline docs/guide/snaps-exports.md:64]:
```suggestion
type onTransactionHandlerReturn = Promise<OnTransactionResponse>;
```

File: docs/guide/snaps-exports.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+```
+
+```typescript
```
Comment (ritave) [inline docs/guide/snaps-exports.md:67]: Why are those separate code blocks?

File: docs/guide/snaps-exports.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+interface OnTransactionResponse {
+  insights: { [key: string]: unknown };
+}
+```
+
+- `onTransactionResponse` - The `insights` object returned by the snap will be displayed alongside the confirmation for the transaction that `onTransaction` was called with. Keys and values will be displayed in the order received, with each key rendered as a title and each value rendered as a string.
```
Comment (ritave) [inline docs/guide/snaps-exports.md:73]: Please add an example section.

File: docs/guide/snaps-permissions.md [hunk 11]
```diff
@@ -29,6 +29,10 @@ For snaps that need to access the internet, the snap can request the `endowment:
 `XMLHttpRequest` is never available in snaps, and you should replace it with `fetch`. If your dependencies are using `XMLHttpRequest`, you can learn how to patch it away [here](./snaps-patching-dependencies.md#patching-the-use-of-xmlhttprequest).
 :::
 
+### `endowment:transaction-insight`
+
+For snaps that provides transaction insights, the snap can request the `endowment:transaction-insight` permission. This permission grants a snap read-only access to raw transaction payloads, before they are accepted for signing by the user, by exporting the `onTransaction` method.
```
Comment (ritave) [inline docs/guide/snaps-permissions.md:34]: Direct to `snap-exports.md` for more details?

**Review by ritave — CHANGES_REQUESTED** [review]

**Review by hmalik88 — DISMISSED** [review]
LGTM.

File: docs/guide/snaps-exports.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+`RpcHandlerReturn` - A promise containing the return of the implemented RPC Method.
+
+### Examples
+
+#### Typescript
+
+```typescript
+import { OnRpcRequestHandler } from '@metamask/snap-types';
+
+export const onRpcRequest: OnRpcRequestHandler = ({ origin, request }) => {
```
Comment (ritave) [inline docs/guide/snaps-exports.md:51]: For consistency with JS version.

```suggestion
export const onRpcRequest: OnRpcRequestHandler = async ({ origin, request }) => {
```

File: docs/guide/snaps-exports.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+interface OnTransactionArgs {
+  transaction: Record<string, unknown>;
+  chainId: string;
+}
+```
+
+### Returns
+
+```typescript
+type onTransactionHandlerReturn = Promise<OnTransactionResponse>;
```
Comment (ritave) [inline docs/guide/snaps-exports.md:98]:
```suggestion
type OnTransactionHandlerReturn = Promise<OnTransactionResponse>;
```

**Review by ritave — APPROVED** [review]

**Review by hmalik88 — DISMISSED** [review]

**Review by ritave — APPROVED** [review]

**Review by ritave — APPROVED** [review]

**Review by ritave — DISMISSED** [review]

**Comment by ritave** [comment]
Wtf. Why can you multiple approve.

File: docs/guide/snaps-permissions.md [hunk 14]
```diff
@@ -29,6 +29,10 @@ For snaps that need to access the internet, the snap can request the `endowment:
 `XMLHttpRequest` is never available in snaps, and you should replace it with `fetch`. If your dependencies are using `XMLHttpRequest`, you can learn how to patch it away [here](./snaps-patching-dependencies.md#patching-the-use-of-xmlhttprequest).
 :::
 
+### `endowment:transaction-insight`
+
+For snaps that provides transaction insights, the snap can request the `endowment:transaction-insight` permission. This permission grants a snap read-only access to raw transaction payloads, before they are accepted for signing by the user, by exporting the `onTransaction` method. see [Exports](./snaps-exports.html#ontransaction)
```
Comment (jlwllmr) [inline docs/guide/snaps-permissions.md:34]: 'provides' -> provide

**Review by jlwllmr — CHANGES_REQUESTED** [review]
A few minor suggestions!

File: docs/guide/snaps-exports.md [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+Do you have feature requests? Other ideas? We'd love to hear about them! [Click here](https://github.com/MetaMask/snaps-skunkworks/discussions) to join the discussion.
+:::
+
+## Table of Contents
+
+[[toc]]
+
+## `onRpcRequest`
+
+In order to communicate with dApps and other snaps, the snap must implement its own JSON-RPC API by exposing an exported function called `onRpcRequest`. Whenever the snap receives a JSON-RPC request, the `onRpcRequest` handler function will be called with the below parameters.
```
Comment (jlwllmr) [inline docs/guide/snaps-exports.md:17]: In line 20 we use 'dapp' (uncapitalised) -- suggest you decide on one for consistency

File: docs/guide/snaps-exports.md [hunk 16]
```diff
@@ -14,10 +14,10 @@ Do you have feature requests? Other ideas? We'd love to hear about them! [Click
...
+In order to communicate with dapps and other snaps, the snap must implement its own JSON-RPC API by exposing an exported function called `onRpcRequest`. Whenever the snap receives a JSON-RPC request, the `onRpcRequest` handler function will be called with the below parameters.
 
 ::: tip Does my snap need to have an RPC API?
-Well, no, that's also up to you! If your snap can do something useful without receiving and responding to JSON-RPC requests, then you can skip exporting onRpcRequest. However, if you want to do something like manage the user's keys for a particular protocol and create a dapp that e.g. sends transactions for that protocol via your snap, you need to specify an RPC API.
+Well, no, that's also up to you! If your snap can do something useful without receiving and responding to JSON-RPC requests, then you can skip exporting onRpcRequest. However, if you want to do something like manage the user's keys for a particular protocol and create a dapp that e.g. sends transactions for that protocol via your snap, for example, you need to specify an RPC API.
```
Comment (jlwllmr) [inline docs/guide/snaps-exports.md:20]: need to remove the 'e.g.' here

**Review by jlwllmr — CHANGES_REQUESTED** [review]
Just one minor thing to amend

**Review by jlwllmr — APPROVED** [review]
👍

**Comment by GuillaumeRx** [comment]
Waiting on the Flask release that will contain transactions insights for Snaps to merge.

**Pull request merged**